### PR TITLE
Add execute governance proposal button and fix voting states

### PR DESCRIPTION
**Pull request by dmanjunath**
My attempt and fixing the Governance voting states and adding an execute button I did yesterday in case we needed to fast track this. Currently what happens in the UI is it only shows you the the submit vote buttons, regardless of whether the proposal is in the voting period, execution delay or is awaiting execution. So not only is there not Execute Proposal button but it still shows you the ability to vote, even after the voting period has passed.

I added substates to "In Progress" so the interface can distinguish these actions and they behave in the following ways:
When voting is allowed, show the voting buttons
After voting has ended but before execution, disable the buttons
After execution is allowed, show the Execute Proposal button

What's left to be done?
Execute Proposal button works, but doesn't show the confirm modal
No "Est Time Remaining" for execution delay
Probably want to display some kind of tag to show current proposal In Progress substatus

During voting period
image

After voting before execution delay
image

During execution delay
image

When eligible to execute
image

### CODE REVIEW

**Comment by raymondjacobson** [comment]
I would have just added these states to the `Outcome` enum. We already have
ApprovedExecuted = 'ApprovedExecuted', // Proposal votingPeriod has closed and vote passed. Proposal was successfully executed.

and that way, the outcome status chip can be 1-1 with these

File: src/components/ProposalHero/ProposalHero.tsx [hunk 1]
```diff
@@ -107,6 +114,83 @@ const VoteCTA: React.FC<VoteCTAProps> = ({
   )
 }
 
+type ExecutionDelayCTAProps = {
+  votingDeadlineBlock: number
+}
+
+const ExecutionDelayCTA: React.FC<ExecutionDelayCTAProps> = ({
```
Comment (dmanjunath) [inline src/components/ProposalHero/ProposalHero.tsx:121]: added different components for each substate, but could possibly combine the execution delay + awaiting execution into a single component with a flag

**Comment by dmanjunath** [comment]
@raymondjacobson finally got around to finishing this. would appreciate any more feedback/best practices you got

File: src/components/ProposalHero/ProposalHero.tsx [hunk 2]
```diff
@@ -107,6 +114,83 @@ const VoteCTA: React.FC<VoteCTAProps> = ({
...
+    ? getActiveStake(accountUser)
+    : Utils.toBN('0')
+  const isUserStaker = userStatus === Status.Success && !activeStake.isZero()
+
+  return (
+    <div className={styles.voteCTA}>
+      <div className={styles.voteStatusContainer}>
+        <div className={styles.title}>{messages.awaitExecuteProposal}</div>
+      </div>
+      {isUserStaker && (
```
Comment (raymondjacobson) [inline src/components/ProposalHero/ProposalHero.tsx:181]: Do you have to be a staker to execute? i would think anyone could hypothetically do this

File: src/components/ProposalHero/ProposalHero.tsx (hunk 1 above)
Comment (raymondjacobson) [reply src/components/ProposalHero/ProposalHero.tsx:121]: eh, yeah i think a prop would be nicer, but this is totally fine too

File: src/store/cache/proposals/hooks.ts [hunk 3]
```diff
@@ -195,15 +212,90 @@ export const useVotingPeriod = () => {
...
+    return null
+
+  if (
+    currentBlockNumber >=
+    submissionBlockNumber + votingPeriod + executionDelay
+  )
+    return Outcome.InProgressAwaitingExecution
+  else if (currentBlockNumber >= submissionBlockNumber + votingPeriod)
+    return Outcome.InProgressExecutionDelay
+  else return Outcome.InProgress
```
Comment (raymondjacobson) [inline src/store/cache/proposals/hooks.ts:276]: nit no else needed

File: src/store/cache/proposals/hooks.ts [hunk 4]
```diff
@@ -195,15 +212,90 @@ export const useVotingPeriod = () => {
...
+}
+
+/**
+ * Calculate the block numbers for major milestones along a proposal's lifetime
+ * submissionBlock - block where the submission occurred
+ * votingDeadlineBlock - block before which all voting must occure
+ * executionDeadlineBlock - block after which a proposal can be executed
+ * @param proposal Proposal object
+ */
+export const useProposalMilestoneBlocks = (proposal: Proposal) => {
```
[the remark this answers is not in the document]
Comment (raymondjacobson) [reply src/store/cache/proposals/hooks.ts:286]: i think this is totally fine

File: src/types.ts [hunk 5]
```diff
@@ -123,6 +123,8 @@ export type VoteEvent = {
 
 export enum Outcome {
   InProgress = 'InProgress', // Proposal is active and can be voted on.
+  InProgressExecutionDelay = 'InProgressExecutionDelay', //Proposal is active but cannot be voted on because it's in execution delay
```
Comment (raymondjacobson) [inline src/types.ts:126]: nit space after //

File: src/components/ProposalStatusBadge/ProposalStatusBadge.tsx [hunk 6]
```diff
@@ -12,6 +12,14 @@ const messages = {
 
 const outcomeMapping = {
   [Outcome.InProgress]: { text: messages.open, style: styles.open },
+  [Outcome.InProgressExecutionDelay]: {
+    text: messages.open,
```
Comment (raymondjacobson) [inline src/components/ProposalStatusBadge/ProposalStatusBadge.tsx:16]: We actually had some loose designs for these terms that didn't make a whole lot of sense, but i'd be down with maybe making this more explicit (check w/ forrest as well) -- something like:

Execution Delay

Awaiting Execution

and changing the color to orange per design for pending states here:
https://www.figma.com/file/csnO7UcYJpF1iu6t0V4F4T/Audius-Protocol-Dashboard?node-id=128%3A0

File: src/store/cache/proposals/hooks.ts [hunk 7]
```diff
@@ -195,15 +212,90 @@ export const useVotingPeriod = () => {
...
+  return { executionDelay }
+}
+
 export const useProposalTimeRemaining = (submissionBlock: number) => {
   const { votingPeriod } = useVotingPeriod()
   const remaining = useTimeRemaining(submissionBlock, votingPeriod)
   return remaining
 }
 
+export const useExecutionDelayTimeRemaining = (votingDeadlineBlock: number) => {
```
Comment (raymondjacobson) [inline src/store/cache/proposals/hooks.ts:233]: should we use these hooks in the <Proposal> list view component?
https://github.com/AudiusProject/protocol-dashboard/blob/main/src/components/Proposal/Proposal.tsx

**Review by raymondjacobson — APPROVED** [review]
Screen Shot 2021-04-23 at 12 20 29 AM

I'd double check w/ forrest on this copy


overall looks great!!!

**Pull request merged**